extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,123 @@
 import com.hazelcast.jet.impl.serialization.SerializerHookConstants;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.ByteArraySerializer;
 import com.hazelcast.nio.serialization.Serializer;
 import com.hazelcast.nio.serialization.SerializerHook;
 import com.hazelcast.nio.serialization.StreamSerializer;
+import org.apache.avro.Schema;
+import org.apache.avro.SchemaNormalization;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
 import org.apache.avro.util.Utf8;
 
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
 
 /**
- * Hazelcast serializer hooks for the classes in the {@code
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private AvroSerializerHooks() { }
 
-    public static final class Utf8Hook implements SerializerHook<Utf8> {
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+            datumWriter.write(data, encoder);
+            encoder.flush();
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericRecordHook implements SerializerHook<GenericRecord> {
+        private final Map<String, Schema> jsonToSchema = new HashMap<>();
```

Comment:
these maps grow unbounded and there is no possibility to clear them

Reply:
This is in line with `MemberSchemaService#schemas` and `ClientSchemaService#schemas`. They even grow bigger because they need to have all schemas in the cluster, not just the ones read/written by the current member. This situation, unfortunately, cannot be easily solved in schema _caches_ since it will require implementing a LRU policy. When we implement a schema _registry_, it will be solved for both Compact and Avro.

Reply:
follow-up comment: https://github.com/hazelcast/hazelcast/pull/25349/files#r1328564425